### PR TITLE
kv/bulk: compress in-memory buffers

### LINKED ISSUE #73983: bulk: use compressed kv buffer in BufferingAdder
Each source node has a B byte buffer in the BufferingAdder for buffering and sorting kv pairs. Assuming a primary index key maps to a random key in the secondary index key space, each buffer flush from a source node will generate keys over the same key space. Let there be N target nodes. Each target node gets B/N of this buffer.
Say N=5 (if we have a 15 node cluster with 3 way replication and the index is evenly distributed across all nodes) and B=512MB. B/N=102.4MB. With 8x compression ratio this is 12.8MB of sstables before repetition of the key space. Increasing the bytes before repetition will make both sstable ingestion (allows for narrower sstables in the key space) and L0 sublevels behave significantly better. For instance if each sst becomes 1/8th of the width in the key space we can fit 8 of these side-by-side in a level, instead of them stacking up from L6 through L0 (with 2 sub-levels in L0).

We can build a compressed kv buffer by constructing and wrapping a temporary in-memory Pebble instance, at some cpu cost in the source.
- Possible settings for this Pebble instance.
  - Disable WAL
  - Set MaxConcurrentCompactions=1 (ideally we should support compaction+flush concurrency of 1 -- this will ensure at most one extra cpu gets used, limiting the effect on foreground traffic).
  - Disable read compactions and read sampling (we see the latter in cpu profiles -- we don't want to waste cpu).
  - Disable deletion pacing (of the in-memory files), so that the sum of the live file sizes + memtable bounds the memory footpring.
  - MemTableStopWritesThreshold=1 so there is at most 1 memtable. We want most of the memory used by compressed files.
  - L0CompactionThreshold=2, L0StopWritesThreshold=3 so we don't build up read amp since we want nearby keys to be in the same sstable for better compression. Alternatively, if key prefix compression is good enough when flushing a single memtable, we could consider tuning the L0 settings such that there is no compaction from L0=>Lbase, even if it means a high read amp. We are going to scan once at the end, and we can afford to do a merge sort across 50 levels for that scan.
- The wrapper would keep track of the uncompressed bytes it is adding. Every N bytes (say 64MB), it would get pebble.Metrics and use the total file size to decide whether to flush. After flushing, this Pebble instance can be discarded.
- There is more memory churn here compared to using bulk.kvBuf. By setting the Pebble options to have the same target file size in each level, and using a customized in-memory FS implementation (not vfs.MemFS), we should be able to reuse the memory allocated for a file when writing a new file.

cc: @dt 

Jira issue: CRDB-11840

Epic CRDB-19402

### PR DESCRIPTION
**Pull request by dt**
Release note: none.

Release justification: high impact change with relatively low risk.

Fixes #73983.

### CODE REVIEW

**Comment by adityamaru** [comment]
Going to give this a review today, sorry for the delay!

File: pkg/kv/bulk/buffering_adder.go [hunk 1]
```diff
@@ -192,49 +209,84 @@ func (b *BufferingAdder) Add(ctx context.Context, key roachpb.Key, value []byte)
...
-}
 
-func (b *BufferingAdder) bufferedKeys() int {
-	return len(b.curBuf.entries)
+	// If we're buffering unsorted data that is not known to compress badly, check
+	// if the buffer is big enough to stop and compress it, both in absolute size
+	// that justifies compression overhead as well as in relative terms to the
+	// size of the budget. The fewer/bigger buffers we have to merge, the cheaper
+	// the merged is, so we want to buffer as much as we can. However, while we're
+	// compressing we will hold both the un-comopressed and compressed in memory
```
Comment (adityamaru) [inline pkg/kv/bulk/buffering_adder.go:244]: nit: s/un-comopressed/un-compressed/g and s/buffing/buffering/g

**Review by adityamaru — APPROVED** [review]
I've done a couple of passes and I have no blocking comments. The changes in buffering_adder.go all make sense to me, as do the new interface and buffers in kv_buf.go. I'm finding it hard to grok the `compressBuffer` and `func (i *compressedBuffer) Next` but that could also just be because it's getting late here. I'm +1 on merging this to give it some bake time but if you think it'd benefit from an additional reviewer I'll let you tag someone in!

**Comment by dt** [comment]
Added some commentary to compressedBuffer's `Next()` and to `compressBuffer` to explain what's going on (but also made it a little more complicated by adding key/key prefix sharing while I was at it.)

**Comment by dt** [comment]
(putting this on ice for a little while)

**Comment by dt** [comment]
@miretskiy @gh-casper after our chat yesterday, this got me thinking if we could perhaps re-use this for our batch wire format

**Comment by miretskiy** [comment]
> @miretskiy @gh-casper after our chat yesterday, this got me thinking if we could perhaps re-use this for our batch wire format

Ack

**Pull request closed without merging**